src/validation.cpp

```diff
@@ -1062,28 +1178,136 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
 
     Workspace ws(ptx);
 
-    if (!PreChecks(args, ws)) return MempoolAcceptResult(ws.m_state);
+    if (!PreChecks(args, ws)) return MempoolAcceptResult::Failure(ws.m_state);
 
     // Only compute the precomputed transaction data if we need to verify
     // scripts (ie, other policy checks pass). We perform the inexpensive
     // checks first and avoid hashing and signature verification unless those
     // checks pass, to mitigate CPU exhaustion denial-of-service attacks.
     PrecomputedTransactionData txdata;
 
-    if (!PolicyScriptChecks(args, ws, txdata)) return MempoolAcceptResult(ws.m_state);
+    if (!PolicyScriptChecks(args, ws, txdata)) return MempoolAcceptResult::Failure(ws.m_state);
 
-    if (!ConsensusScriptChecks(args, ws, txdata)) return MempoolAcceptResult(ws.m_state);
+    if (!ConsensusScriptChecks(args, ws, txdata)) return MempoolAcceptResult::Failure(ws.m_state);
 
     // Tx was accepted, but not added
     if (args.m_test_accept) {
-        return MempoolAcceptResult(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+        return MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions), ws.m_base_fees);
     }
 
-    if (!Finalize(args, ws)) return MempoolAcceptResult(ws.m_state);
+    if (!Finalize(args, ws)) return MempoolAcceptResult::Failure(ws.m_state);
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return MempoolAcceptResult(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+    return MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+}
+
+PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::vector<CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+
+    PackageValidationState package_state;
+    const unsigned int package_count = txns.size();
+
+    // These context-free package limits can be checked before taking the mempool lock.
+    if (package_count > MAX_PACKAGE_COUNT) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-many-transactions");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    const int64_t total_size = std::accumulate(txns.cbegin(), txns.cend(), 0,
+                               [](int64_t sum, const auto& tx) { return sum + GetVirtualTransactionSize(*tx); });
+    // If the package only contains 1 tx, it's better to report the policy violation on individual tx size.
+    if (package_count > 1 && total_size > MAX_PACKAGE_SIZE * 1000) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-large");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    std::vector<Workspace> workspaces{};
+    workspaces.reserve(package_count);
+    {
+        std::unordered_map<uint256, bool, SaltedTxidHasher> txids_seen;
```

Comment:
You don't need a map here, you can just use a set and erase as you go:

<details>
<summary>Diff</summary>

```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 0b429adfae..c72b5eee02 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1226,9 +1226,9 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
     std::vector<Workspace> workspaces{};
     workspaces.reserve(package_count);
     {
-        std::unordered_map<uint256, bool, SaltedTxidHasher> txids_seen;
-        std::transform(txns.cbegin(), txns.cend(), std::inserter(txids_seen, txids_seen.end()),
-                       [](const auto& tx){ return std::make_pair(tx->GetHash(), false); });
+        std::unordered_set<uint256, SaltedTxidHasher> later_txs;
+        std::transform(txns.cbegin(), txns.cend(), std::inserter(later_txs, later_txs.end()),
+                       [](const auto& tx){ return tx->GetHash(); });
 
         // Require the package to be sorted in order of dependency, i.e. parents appear before children.
         // An unsorted package will fail anyway on missing-inputs, but it's better to quit earlier and
@@ -1236,18 +1236,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
         // spend nonexistent coins).
         for (const auto& tx : txns) {
             for (const auto& input : tx->vin) {
-                auto it = txids_seen.find(input.prevout.hash);
-                if (it == txids_seen.end()) {
-                    // The parent is not in this package.
-                    continue;
-                } else if (!it->second) {
+                auto it = later_txs.find(input.prevout.hash);
+                if (it != later_txs.end()) {
                     // The parent is a subsequent transaction in the package.
                     package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-not-sorted");
                     return PackageMempoolAcceptResult(package_state, {});
                 }
             }
-            // Mark this transaction as seen.
-            txids_seen[tx->GetHash()] = true;
+            // Remove from later_txs
+            later_txs.erase(tx->GetHash());
             workspaces.emplace_back(Workspace(tx));
         }
     }
```

</details>

Reply:
Or you could reverse iterate and add as you go:

<details>
<summary>CLICK HERE FOR CODE!</summary>

```c++
    {
        std::unordered_set<uint256, SaltedTxidHasher> later_txs;

        // Require the package to be sorted in order of dependency, i.e. parents appear before children.
        // An unsorted package will fail anyway on missing-inputs, but it's better to quit earlier and
        // fail on something less ambiguous (missing-inputs could also be an orphan or trying to
        // spend nonexistent coins).
        //
        // Iterate through the package backwards, adding the txid to later_txs. If an earlier tx spends
        // an output from later_txs, the package is mis-sorted.
        for (auto txit = txns.crbegin(); txit != txns.crend(); ++txit) {
            for (const auto& input : (*txit)->vin) {
                if (auto it = later_txs.find(input.prevout.hash); it != later_txs.end()) {
                    // The parent is a later transaction in the package.
                    package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-not-sorted");
                    return PackageMempoolAcceptResult(package_state, {});
                }
            }
            // Add to later_txs
            later_txs.insert((*txit)->GetHash());
        }
    }

    std::vector<Workspace> workspaces{};
    workspaces.reserve(package_count);

    std::transform(txns.cbegin(), txns.cend(), std::back_inserter(workspaces), [](const auto& tx) {
        return Workspace(tx);
    });
```

(note that we can't add to workspace as we iterate backwards, because otherwise it'd be reversed)

</details>

Reply:
Clever!

I guess the only reason to prefer the bool version is if we might handle in-package vs out of package parents differently in the future. Absent such a future need, I agree this works nicely!

The reverse iteration version should be a fair bit more efficient because we only need to hash each tx once for the set lookup, rather than 1x to create and 1x to remove, although it's a bit more complex.

Reply:
Noice 🧠 I've done it this way, it's prettier